src/Fantomas.Core/CodePrinter.fs

```diff
@@ -2309,48 +2314,52 @@ let genFunctionNameWithMultilineLids (trailing: Context -> Context) (longIdent:
         )
     | _ -> sepNone
 
-let (|EndsWithDualListApp|_|) (config: FormatConfig) (appNode: ExprAppNode) =
-    if not (config.ExperimentalElmish || config.IsStroustrupStyle) then
-        None
-    else
-        let mutable otherArgs = ListCollector<Expr>()
-
-        let rec visit (args: Expr list) =
-            match args with
-            | [] -> None
-            | [ Expr.ArrayOrList firstList; Expr.ArrayOrList lastList ] -> Some(otherArgs.Close(), firstList, lastList)
-            | arg :: args ->
-                otherArgs.Add(arg)
-                visit args
-
-        visit appNode.Arguments
+[<RequireQualifiedAccess>]
+type AppNode =
+    /// Example: div [ ClassName "container" ] [ str "meh" ]
+    | EndsWithDualList of otherArguments: Expr list * firstList: ExprArrayOrListNode * lastList: ExprArrayOrListNode
+    /// Example: input [ Type "text" ]
+    | EndsWithSingleList of otherArguments: Expr list * list: ExprArrayOrListNode
+    /// Example: a {| X = y |}
+    | EndsWithSingleRecord of otherArguments: Expr list * record: Expr
+    /// Example: Foo(a,b)
+    | SingleParenArgument of appNode: ExprAppNode * parenthesesArgument: Expr
+    /// Example: path.Replace("../../../", "....")
+    | LongIdentAndSingleParenArgument of longIdent: ExprOptVarNode * parenthesesArgument: Expr
+    /// Example: myFunc a b c
+    | Regular of appNode: ExprAppNode
+
+// TODO: memoize function based on the range of appNode
```

Comment:
Is memoizing something you'd like specific to this task, or is that a general improvement you'd like to make at some point?

Reply:
Specifically in this case.
During the pipeline of `CodePrinter` on multiple occasions, we need to transform `ExprAppNode` to `AppNode`. This transformation is always going to be the same (in a single run), so it should not be recalculated each time we invoke the function.